zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/monitoring/HealthTreeMetrics.java

```diff
@@ -0,0 +1,129 @@
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class HealthTreeMetrics implements ComponentTreeListener {
+  public static final String NODES_NAME = "zeebe.broker.health.nodes";
+  private static final Logger LOG = LoggerFactory.getLogger(HealthTreeMetrics.class);
+  private static final int MAX_PATH_ITERATIONS = 8;
+  private final MeterRegistry meterRegistry;
+  private final Map<String, Meter.Id> meters = new HashMap<>();
+  // Map of child -> parent: a child only has 1 parent
+  private final Map<String, String> relationships = new HashMap<>();
+  private final Tag extraTags;
+
+  public HealthTreeMetrics(final MeterRegistry meterRegistry) {
+    this(meterRegistry, null);
+  }
+
+  public HealthTreeMetrics(final MeterRegistry meterRegistry, final Tag extraTags) {
+    this.meterRegistry = meterRegistry;
+    this.extraTags = extraTags;
+  }
+
+  /**
+   * @param status to be mapped
+   * @return a numerical value mapped from status that can be reported by the gauge
+   */
+  private static int statusValue(final HealthStatus status) {
+    return switch (status) {
+      case HEALTHY -> 1;
+      case UNHEALTHY -> -1;
+      case DEAD -> -2;
+      case null -> 0;
+    };
+  }
+
+  @Override
+  public void registerNode(final HealthMonitorable component) {
+    final var gauge = buildNodeGauge(component);
+    meters.put(component.componentName(), gauge.getId());
+  }
+
+  @Override
+  public void unregisterNode(final HealthMonitorable component) {
+    removeFromRegistry(component.componentName());
+  }
+
+  @Override
+  public void registerRelationship(final String child, final String parent) {
+    relationships.put(child, parent);
+  }
+
+  @Override
+  public void unregisterRelationship(final String child, final String to) {
+    final var parent = relationships.get(child);
+    if (parent == null || !parent.equals(to)) {
+      LOG.warn("Tried to remove invalid relationship: child={}, parent={}", child, to);
+    } else {
+      relationships.remove(child);
+    }
+  }
+
+  @Override
+  public void close() {
+    for (final var id : meters.values()) {
+      meterRegistry.remove(id);
+    }
+    meters.clear();
+  }
+
+  private Gauge buildNodeGauge(final HealthMonitorable component) {
+    return Gauge.builder(
+            NODES_NAME,
+            component,
+            // make sure to not close over anything else than `c`
+            c -> c != null ? statusValue(c.getHealthReport().getStatus()) : statusValue(null))
+        .tags(
+            tagsFor(
+                Tags.of(
+                    "id", component.componentName(), "path", pathOf(component.componentName()))))
+        .register(meterRegistry);
+  }
+
+  public String pathOf(final String from) {
+    final var pathElements = new LinkedList<String>();
+    pathElements.add(from);
+    var parent = relationships.get(from);
+    var i = 0;
+    while (parent != null && i++ < MAX_PATH_ITERATIONS) {
```

Comment:
:thought_balloon: Nice foresight, but unless we have a bug, I doubt we'll reach that...though famous last words right? :smile:

Reply:
Yes, since it's just to prevent something very bad because of a mistake in a "secondary" feature.